File: src/game/board.ts
import { findNewCaptures, pointInsideAnyCapture, scoreCaptures } from "./capture";
import type { GameState, Player, Point, Stone } from "./types";

export const pointKey = ({ x, y }: Point): string => `${x}:${y}`;

export const createGameState = (): GameState => ({
  currentPlayer: "red",
  stones: new Map<string, Stone>(),
  captures: [],
  score: { red: 0, blue: 0 }
});

export const otherPlayer = (player: Player): Player => (player === "red" ? "blue" : "red");

export const placeStone = (state: GameState, point: Point): GameState => {
  const key = pointKey(point);
  if (state.stones.has(key) || pointInsideAnyCapture(point, state.captures)) return state;

  const player = state.currentPlayer;
  const stones = new Map(state.stones);
  stones.set(key, { ...point, player });

  const captures = [...state.captures, ...findNewCaptures(stones, state.captures, player, point)];

  return {
    ...state,
    stones,
    captures,
    score: scoreCaptures(captures),
    currentPlayer: otherPlayer(player)
  };
};

File: crates/game-core/src/board.rs
use crate::capture::{point_inside_any_capture, resolve_captures_after_move, score_captures};
use crate::types::{GameState, Point, Stone};

pub fn create_game_state() -> GameState {
    GameState::new()
}

pub fn place_stone(state: &GameState, point: Point) -> GameState {
    if !point.is_safe() || state.has_stone(point) || point_inside_any_capture(point, &state.captures) {
        return state.clone();
    }

    let player = state.current_player;
    let mut next = state.clone();
    next.push_stone(Stone {
        x: point.x,
        y: point.y,
        player,
    });
    next.captures = resolve_captures_after_move(&next, &state.captures, player, point);
    next.score = score_captures(&next.captures);
    next.current_player = player.other();
    next
}
